client/src/classes/PeerProcessor.ts

```diff
@@ -0,0 +1,233 @@
+import { UserInfo, defaultUserInfo } from "../contexts/UserInfoContext";
+import { AudioVisualizer, gainToMultiplier } from "./AudioVisualizer";
+import { Socket } from "socket.io-client";
+import { timeout, DCLabel, Pack } from "./Network";
+
+export interface DataPackage {
+  position: [number, number];
+  info: UserInfo;
+}
+
+export class PeerProcessor {
+  peerConnection: RTCPeerConnection;
+  dataChannel: RTCDataChannel;
+  socket: Socket;
+  sender: RTCRtpSender;
+  peerId: string;
+  stream: MediaStream;
+  player: HTMLAudioElement;
+  visualizer: AudioVisualizer;
+  multiplier: number;
+  // a function to update the positions array context
+  addUserInfo: (info) => void;
+  selfUserInfo: UserInfo;
+  peerUserInfo: UserInfo;
+  constructor(peerId: string, socket: Socket, addUserInfo: (info) => void) {
+    this.peerId = peerId;
+    this.sender = null as unknown as RTCRtpSender;
+    this.dataChannel = null as unknown as RTCDataChannel;
+    // initialize with a free public STUN server to find out public ip, NAT type, and internet side port
+    this.peerConnection = new RTCPeerConnection({
+      iceServers: [{ urls: "stun:stun.l.google.com:19302" }],
+    });
+    this.multiplier = 0;
+    this.stream = new MediaStream();
+    this.visualizer = null as unknown as AudioVisualizer;
+    this.player = new Audio();
+    // the function is re-assigned during the user's initialization
+    this.addUserInfo = addUserInfo;
+    this.selfUserInfo = defaultUserInfo;
+    this.peerUserInfo = defaultUserInfo;
+    this.socket = socket;
+
+    // listener for when a peer adds a track
+    this.peerConnection.ontrack = (event) => {
+      this.updateLocalTrack(event.track);
+    };
+
+    this.peerConnection.onnegotiationneeded = () => {
+      this.sendOffer();
+    };
```

Comment:
When does the peerconnection needs an negotiation? Multiple offer/answer might happen because they need a negotiation??

Reply:
![Selection_093](https://user-images.githubusercontent.com/9146455/134226073-8e81d7d1-cf85-4a31-88fb-c110fae3ae98.png)

Reply:
After iceRestart()